File: vgrep.py
# Vagn Johansen (2017)

import sys
import os
import glob
import argparse
import re
import subprocess
import chardet

parser = argparse.ArgumentParser(description='Grep directories.')
parser.add_argument('-e', default=':code')
parser.add_argument('-m')
parser.add_argument('-A', type=int)
parser.add_argument('-B', type=int)
parser.add_argument('-i', action='store_const', const='i', default='')
parser.add_argument('-r', action='append')
parser.add_argument('dirs', nargs='*')

args = parser.parse_args()
word = args.dirs.pop(0)

exts = re.sub(':code','c,cpp,cc,cxx,cs,h,hpp,hh,asm,el,pl,pm,js,py,ml,cs,java' +
              ',cls,bas,pas,frm,sh,zsh,rb,php,ts,fs,fsx,r,m,xaml' +
              ',ts,tsx,jsx,json',args.e)
exts = re.sub(':text','org,txt,log,htm,html,mak,csproj,sln,vcproj,proj,bat,zsh,' +
              'config,xslt,xsl,css,asp,xml,xsl,xslt,sql', exts)

ext_re = '|'.join(exts.split(','))
ext_re = '\.('+ext_re+')$'
exitcode = 1

homepath = None
if os.getenv("HOMEPATH"):
    homepath = (os.getenv("HOMEDRIVE") + os.getenv("HOMEPATH"))

def grep(dir_):
    if dir_ == 'NUL':
        return 0
    dir_ = dir_.replace('~', os.getenv("HOME") or homepath or "~")
    if not os.path.isdir(dir_):
        print('Directory does not exists:', dir_)
        return 0
    global exitcode
    count = 0
    os.chdir(dir_)
    filenames = list(filter(lambda f: re.search(ext_re,f.lower()), glob.glob('*.*')))
    if len(filenames) == 0:
        return 0
    count += len(filenames)
    shell_args = ['grep','-{}nH'.format(args.i), word]
    if args.A:
        shell_args.append('-A')
        shell_args.append(str(args.A))
    if args.B:
        shell_args.append('-B')
        shell_args.append(str(args.A))
    filenames = list(map(lambda fn: fn.replace('{','\{').replace('}','\}'), filenames))
    # filenames = [f for f in filenames if f != 'main.js']
    shell_args.extend(filenames)
    sys.stdout.flush()
    try:
        rawdata = subprocess.check_output(shell_args, shell=False)
        output = rawdata.decode('ascii', 'ignore')
    except subprocess.CalledProcessError:
        return count
    except UnicodeEncodeError as ex:
        print('UnicodeEncodeError Exception', dir_, ex)
    except UnicodeDecodeError as ex:
        print('UnicodeDecodeError Exception', dir_, ex)
        return count

    # python 3.5 can use
    #    output = subprocess.run(shell_args, stdout=subprocess.PIPE).stdout.decode('utf-8')
    if len(output) > 2:
        exitcode = 0
        print('Entering directory `{}\''.format(dir_))
        print(output.rstrip())
        print('Leaving directory `{}\''.format(dir_))
        print()
    else:
        print('No matches in',dir_)

    return count

count = 0
for dir_ in args.dirs:
    count += grep(dir_)

if args.r is not None:
    for rdir in args.r:
        rdir = rdir.replace("%20"," ")
        for root, dirs, files in os.walk(rdir, topdown=True):
            dirs[:] = [d for d in dirs if d != '.git' and d != 'node_modules' and d != 'build' and d != 'dist']
            count += grep(root)

print('{} files searched'.format(count))
sys.exit(exitcode)

File: vgrep.mjs
#!/usr/bin/env node
// Vagn Johansen (2017) - ported to Node.js 2025

import fs from 'fs';
import path from 'path';
import { spawnSync } from 'child_process';

function parseArgs() {
  const args = process.argv.slice(2);
  let opts = {
    e: ':code',
    m: null,
    A: null,
    B: null,
    i: false,
    r: [],
    dirs: [],
  };
  let i = 0;
  while (i < args.length) {
    let arg = args[i];
    if (arg === '-e') {
      opts.e = args[++i];
    } else if (arg === '-m') {
      opts.m = args[++i];
    } else if (arg === '-A') {
      opts.A = parseInt(args[++i], 10);
    } else if (arg === '-B') {
      opts.B = parseInt(args[++i], 10);
    } else if (arg === '-i') {
      opts.i = true;
    } else if (arg === '-r') {
      opts.r.push(args[++i]);
    } else if (arg.startsWith('-')) {
      // ignore unknown
    } else {
      opts.dirs.push(arg);
    }
    i++;
  }
  return opts;
}

const args = parseArgs();
if (args.dirs.length === 0) {
  console.error('Usage: node vgrep.js <word> [dirs...]');
  process.exit(1);
}
const word = args.dirs.shift();

let exts = args.e.replace(
  /:code/,
  'c,cpp,cc,cxx,cs,h,hpp,hh,asm,el,pl,pm,js,py,ml,cs,java' +
    ',cls,bas,pas,frm,sh,zsh,rb,php,ts,fs,fsx,r,m,xaml' +
    ',ts,tsx,jsx,json',
);
exts = exts.replace(
  /:text/,
  'org,txt,log,htm,html,mak,csproj,sln,vcproj,proj,bat,zsh,' +
    'config,xslt,xsl,css,asp,xml,xsl,xslt,sql',
);

const ext_re = new RegExp('\\.(' + exts.split(',').join('|') + ')$', 'i');
let exitcode = 1;

const homepath = process.env.HOMEPATH
  ? process.env.HOMEDRIVE + process.env.HOMEPATH
  : null;

function grep(dir_) {
  if (dir_ === 'NUL') return 0;
  dir_ = dir_.replace('~', process.env.HOME || homepath || '~');
  if (!fs.existsSync(dir_) || !fs.statSync(dir_).isDirectory()) {
    console.log('Directory does not exists:', dir_);
    return 0;
  }
  let count = 0;
  const files = fs.readdirSync(dir_).filter((f) => ext_re.test(f.toLowerCase()));
  if (files.length === 0) return 0;
  count += files.length;
  let shell_args = ['-nH'];
  if (args.i) shell_args.unshift('-i');
  if (args.A) {
    shell_args.push('-A');
    shell_args.push(String(args.A));
  }
  if (args.B) {
    shell_args.push('-B');
    shell_args.push(String(args.B));
  }
  shell_args.push(word);
  shell_args = shell_args.concat(
    files.map((fn) => fn.replace('{', '\\{').replace('}', '\\}')),
  );
  const grepProc = spawnSync('grep', shell_args, { cwd: dir_, encoding: 'utf8' });
  let output = grepProc.stdout || '';
  if (grepProc.error) {
    // grep not found or failed
    return count;
  }
  if (output.length > 2) {
    exitcode = 0;
    console.log(`Entering directory \`${dir_}'`);
    process.stdout.write(output.trimEnd() + '\n');
    console.log(`Leaving directory \`${dir_}'\n`);
  } else {
    console.log('No matches in', dir_);
  }
  return count;
}

let count = 0;
for (const dir_ of args.dirs) {
  count += grep(dir_);
}

if (args.r && args.r.length > 0) {
  for (let rdir of args.r) {
    rdir = rdir.replace(/%20/g, ' ');
    function walk(root) {
      let subdirs;
      try {
        subdirs = fs.readdirSync(root, { withFileTypes: true });
      } catch (e) {
        return;
      }
      let dirs = [];
      for (const entry of subdirs) {
        if (
          entry.isDirectory() &&
          !['.git', 'node_modules', 'build', 'dist'].includes(entry.name)
        ) {
          dirs.push(entry.name);
        }
      }
      count += grep(root);
      for (const d of dirs) {
        walk(path.join(root, d));
      }
    }
    walk(rdir);
  }
}

console.log(`${count} files searched`);
process.exit(exitcode);
